pennylane/devices/qutrit_mixed/measure.py

```diff
@@ -0,0 +1,308 @@
+# Copyright 2018-2024 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+Code relevant for performing measurements on a qutrit mixed state.
+"""
+
+from typing import Callable
+from string import ascii_letters as alphabet
+from pennylane import math
+from pennylane.ops import Sum, Hamiltonian
+from pennylane.measurements import (
+    StateMeasurement,
+    MeasurementProcess,
+    ExpectationMP,
+    StateMP,
+    ProbabilityMP,
+    VarianceMP,
+)
+from pennylane.operation import Observable
+from pennylane.typing import TensorLike
+
+from .utils import (
+    get_einsum_mapping,
+    resquare_state,
+    get_num_wires,
+    get_new_state_einsum_indices,
+    QUDIT_DIM,
+)
+from .apply_operation import apply_operation
+
+alphabet_array = math.asarray(list(alphabet))
+
+
+def apply_observable_einsum(obs: Observable, state, is_state_batched: bool = False):
```

Comment:
How does this differ from `apply_operation_einsum`?

Reply:
Gabe can elaborate more but `apply_operation_einsum` handles batches of Kraus operators and thus operation on the state from both sides. I guess these could be merged/streamlined by redefining this one generically in `utils`, to just perform the matrix multiplication, and then using it in this module straight up (and adjusting the `apply_operation` one to apply it twice.

Reply:
Yes

1. `apply_operation_einsum` takes operation (with matrix $A$) and state with density martix $\rho$ and returns $A\rho A^\dagger$ and in the case of a channel it returns $\sum_i K_i\rho K_i^\dagger$. where K is the set of Kraus matrices
2. `apply_observable_einsum` takes observable (with matrix $O$) and state with density martix $\rho$ and returns $O\rho$ this is used to take the trace for expectation value $<O>=Tr(O\rho)$

I could definitly implement it as @glassnotes suggests, but I think it would actually complicate things and make `apply_operation` slower as it currently uses the equivalent of:
```math.einsum(f"...{kraus_indices},...{state_indices},...{kraus_dag_indices}->...{new_state_indices}", kraus, state, kraus_dagger)```
I think by separating this would require an equivalent amount of repeated code.

It may make more sense to instead require diagonalizing gates for all measurements and not include the trace method if this repetition is not ideal

Reply:
Thanks for the explanation 👍

Reply:
Makes sense, thanks @Gabriel-Bottrill !